### PR TITLE
Add `calculate_remove_liquidity`

### PR DESCRIPTION
**Pull request by sentilesdal**
# Resolved Issues
https://github.com/delvtech/hyperdrive-rs/issues/22

# Description
Adds `calculate_remove_liquidity` and a test.

# Review Checklists

Please check each item **before approving** the pull request. While going
through the checklist, it is recommended to leave comments on items that are
referenced in the checklist to make sure that they are reviewed.

- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?
    - [ ] If matching Solidity behavior, are there differential fuzz tests that
          ensure that Rust matches Solidity?

### CODE REVIEW

**Comment by dpaiton** [comment]
would you mind adding the python bindings?

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (dpaiton) [inline crates/hyperdrive-math/src/lp/remove.rs:34]: you can declare the variable as `mut` in the fn signature and then use `+=`. I am not sure which is better, tbh, but the latter pattern is what we use e.g. in [yieldspace.rs](https://github.com/delvtech/hyperdrive-rs/blob/dpaiton/calc-open-short/crates/hyperdrive-math/src/yield_space.rs#L274) and [short/open.rs](https://github.com/delvtech/hyperdrive-rs/blob/dpaiton/calc-open-short/crates/hyperdrive-math/src/short/open.rs#L43).

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (dpaiton) [inline crates/hyperdrive-math/src/lp/remove.rs:47]: Elsewhere (e.g. [short/open](https://github.com/delvtech/hyperdrive-rs/blob/dpaiton/calc-open-short/crates/hyperdrive-math/src/short/open.rs#L160-L177) and [long/open](https://github.com/delvtech/hyperdrive-rs/blob/dpaiton/calc-open-short/crates/hyperdrive-math/src/long/open.rs#L53-L76)) we have been separating functions that compute the new state and new deltas. This case is a little tricky... You permute state as part of the process (adjusting lp total supply), which I assume is then considered when calculating redeem withdrawal shares.

Do you think it makes sense to include the new `state` in the return tuple, or you could return a `lp_supply_delta` variable to be used in a later `calculate_pool_state_after_remove_liquidity` function. Or don't add an extra variable and re-compute it in a dedicated `calculate_pool_deltas*` fn.

Thoughts?

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (dpaiton) [inline crates/hyperdrive-math/src/lp/remove.rs:91]: solidity burns shares from the user here; was that skipped on purpose?

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (dpaiton) [inline crates/hyperdrive-math/src/lp/remove.rs:137]: solidity has `_distributeExcessIdleSafe()`, which I believe does the actual distribution & includes checks, and `calculateDistributeExcessIdle()` to compute the `withdrawalSharesRedeemed` and `shareProceeds`.

Here in rust you're renaming the first function to `calculate_distribute_excess_idle_safe` and the second to `calculate_distribute_excess_idle`. I think these names are confusing -- it implies that they're the same function and one is the "safe" version of it. 

This is a case where we should depart from exactly copying solidity in favor of rust readability. I think you should rename one of these functions.

**Review by dpaiton — APPROVED** [review]
lgtm but @jalextowle  said he was going to review as well so you should wait for confirmation from him before merging

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp/remove.rs:34]: It's been my preference to not use `mut` when possible and just redeclare.  Avoiding mutability is kinda Rust's bag.

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp/remove.rs:91]: skipped on purpose, rust doesn't have that state.  these are `calculate_x` functions so as long as we return a correct calculation we can ignore resulting state updates.  Also, by state I mean contract state, i.e. things like user token balances, not the `State` object we pass around.

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (jalextowle) [inline crates/hyperdrive-math/src/lp/remove.rs:290]: I'm not sure why we need this. Shouldn't this be perfectly accurate?

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (jalextowle) [inline crates/hyperdrive-math/src/lp/remove.rs:299]: Same here

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (dpaiton) [reply crates/hyperdrive-math/src/lp/remove.rs:37]: oh ok, I didn't know that. Should we change those examples I gave you then?

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (dpaiton) [reply crates/hyperdrive-math/src/lp/remove.rs:100]: ok cool thanks for clarifying :)

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp/remove.rs]: this is removed, exact now!

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp/remove.rs]: exact now!

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp/remove.rs:37]: I wouldn't go out of your way but in general we should avoid mutability unless needed for a for-loop scenario etc.

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp/remove.rs:129]: It seems like the pattern is to call things 'calculate' because we aren't updating any state, just returning values.

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp/remove.rs:53]: If there's a need for it we can, until then I'd say leave as is?

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp/remove.rs:153]: > This is a case where we should depart from exactly copying solidity in favor of rust readability. I think you should rename one of these functions.

This was my feeling as well, but honestly it's too much of a pain to maintain two different implementations for too little gain.  I tore my hair out trying to make things more readable only to revert to a more mirrored approach.  

I can rename this to whatever we want though, it's 'safe' because of the extra checks, how about renaming to get rid of the 'calculate' part and return updated `State` along with the updated values?

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp/remove.rs:129]: Ok actually yeah I'll remove the 'calculate' part and return updated state as a part of this function.

Comment (jalextowle) [reply crates/hyperdrive-math/src/lp/remove.rs:129]: Ok sounds good. I'll do an early approve, but make sure you address this before merging.

**Review by jalextowle — APPROVED** [review]
LGTM

File: crates/hyperdrive-math/src/lp/remove.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp/remove.rs]: Ok I actually added this in because it helped with some extra test scenarios.

**Pull request merged**
